User/savoyschuler/pagercontrol (#88)
microsoft/microsoft-ui-xaml-specs: active/PagerControl/PagerControl.md

### Localization
* The numbers will follow standard behavior when localized to/from a language that uses a numeral system that differs from Western Arabic. 
* The prefix and suffix text will follow standard mirroring behavior when localized to/from languages that are RTL or LTR. 

## Data and Intellegence Metrics 

Adoption and validation of the control in the community
* Measurement: Use telemetry to determine how many first and third party developers are using the control in preview. 
* Measurement: Count of posts providing feedback to us through Github and Discord. 

Questions that can be answered through telemetry metrics to help focus future features and improvements: 
Which display mode is the most commonly used? 
* Measurement: Count of how many applications include the pager control type broken out by display mode.

How often are developers using this control with an indefinite number of pages?
* Measurement: Count of how many applications include the pager control type and have the NumberOfPages property empty.

## Open Questions 

> * How does narrator handle localization? In the “page x of y” scenario how would that work if we switch from LTR to RTL?

Reviewer: Does RTL or LTR make a difference in this scenario? Isn't that something that we are providing through the AutomationPeer which provides the PositionInSet and SizeOfSet values to the narrator?

Reviewer 2: I don't think anything special is needed here for RTL, but @MikeHillberg can say for sure. If the items get flipped, the elements would still have the correct position/size properties still.

Author: Yes but I am not sure if narrator will correctly read page x of y in that case if the localized language is RTL. @StephenLPeters mentioned this before and I wanted to add it as an open question to make sure I talk with accessibility about it.

Author: @chingucoding you are right this is not an issue since we are using the PositionInSet property. Good catch!

## Examples
The pager control has three different display modes and each display mode has configurable components within them. The examples below show the default configurations for each mode. 

### ComboBox Display Mode
The Prefix and Suffix properties are specific to the combo box and number box modes. They will be ignored if the button panel mode is chosen as the display. 

![](images/pager-control-combobox.png)

XAML
```XAML
<GridView x:Name="gridView1" ... />
<controls:PagerControl x:Name="MainPagerControl"
        DisplayMode="ComboBox"
	NumberOfPages="10"
        FirstButtonVisibility="None"
        PreviousButtonVisibility="AlwaysVisible"
        PreviousButtonText="Prev"
        NextButtonVisibility="AlwaysVisible"
        NextButtonText="Next"
        LastButtonVisibility="None"
	FirstButtonCommand="FirstButtonPressedEvent"
        PreviousButtonCommand="PreviousButtonPressedEvent"
        NextButtonCommand="NextButtonPressedEvent"
        LastButtonCommand="LastButtonPressedEvent"
        PagerInputCommand="PagerInputEvent"
	Prefix="Page"

> 	Suffix="of 10"/>

Reviewer: nit: Indentation

</Grid>
```

### Pager Control Button Panel Display Mode
The EllipsisAreEnabled and EllipsisShowFirstAndLast properties are specific to this mode. They will be ignored if another mode is selected.

![](images/pager-control-numerical-button-panel.png)

XAML
```XAML
<GridView x:Name="gridView1" ... />
<controls:PagerControl x:Name="MainPagerControl"
        DisplayMode="NumericalButtonPanel"

> 	NumberOfPages="10"

Reviewer: nit: Indentation

        FirstButtonVisibility="None"
        PreviousButtonVisibility="AlwaysVisible"
        NextButtonVisibility="AlwaysVisible"
        LastButtonVisibility="None"
        FirstButtonCommand="FirstButtonPressedEvent"
        PreviousButtonCommand="PreviousButtonPressedEvent"
        NextButtonCommand="NextButtonPressedEvent"
        LastButtonCommand="LastButtonPressedEvent"
        PagerInputCommand="PagerInputEvent"
        EllipsisAreEnabled="True"
	EllipsisShowFirstAndLast="True"
	/>
</Grid>
```

### Pager Control Editable NumberBox Display Mode 
The Prefix and Suffix properties are specific to the combo box and number box modes. They will be ignored if the button panel mode is selected.

![](images/pager-control-editable-textbox.png)

XAML
```XAML
<GridView x:Name="gridView1" ... />
<controls:PagerControl x:Name="MainPagerControl"
        DisplayMode="NumberBox"
        NumberOfPages="5"
	FirstButtonVisibility="None"
        PreviousButtonVisibility="AlwaysVisible"
        NextButtonVisibility="AlwaysVisible"
        LastButtonVisibility="None"

> 	FirstButtonCommand="FirstButtonPressedEvent"

Reviewer: nit: indentation of FirstButtonVisibility and FirstButtonCommand

| PrefixText | * Text displayed before the editable ComboBox or NumberBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| SuffixText | * When a total number of indices (N) is given, this suffix string will appear after the editable ComboBox indexing component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 
enum PagerDisplayMode
{
    Auto,
    ComboBox,
    NumberBox,
    ButtonPanel,
};

enum PagerButtonVisibilityBehavior
{
    Auto,
    AlwaysVisible,
    HiddenOnEdge,
    None,
};

runtimeclass PagerControlPageChangedEventArgs
{

>     Integer CurrentPage{get; };

Reviewer: I am not sure if CurrentPage is the best name here, especially depending on when the event get's raised. On the other hand, NavigationView uses SelectedItem, so we might be fine here. A potential issue could be that we might raise the event in a stage where CurrentPage of the args and the actual current page might be different.

Author: I updated it to be "NewIndex" but I am not convinced New is the best name.

Button Panel
* Max wants to search a clothing store's inventory to find a new pair of blue tennis shoes. He searches for blue tennis shoes on the site and is presented with the list of search results. He then scans through the images and descriptions on the first page and does not find anything he likes. He then navigates to the next page and continues searching until he finds the blue tennis shoes he is looking for. 

Combo Box/Number Box
* Jill opens an application that displays the produce inventory for her grocery store. She sorts her list of results alphabetically and wants to navigate to the page that has the results starting with "L". She knows that what she is looking for is on one of the pages in the middle and wants to navigate there from the list of pages. From there, she can go to the next or previous page to find where the items starting with "L" are. 

Any Display Mode
* Samanatha opens up an application for the first time and is presented with a tutorial window to show her how to use the application. She needs to move through each page of the tutorial sequentially in order to continue using the application. 

## Accessibility

### UI Automation Patterns 

> * ComboBox and NumberBox will use the control patterns already assigned to them. The pager control will use the selection provider pattern and the individual buttons will use selection item provider. 

Reviewer: Will we use selection provider for all three kinds (ComboBox, NumberBox and buttons)? How would selection through UIA work then?

Author: I am going back and forth between using invoke or selection for the control but you bring up a good point about selection and the other UIA patterns for ComboBox and NumberBox and the UIA patterns they already have. Since everything outside of NumberBox and ComboBox is a button it might make more sense to follow the invoke pattern.

# Pager control

## Background

A UI pager is a common control that lets a user page through content, rather than scroll or pan. A common experience is to provide preview/next buttons, and a way to jump to a specific page.

> The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this control is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.

Reviewer 3: Formatting note: hard returns in markdown (unfortunately) makes PRs easier, because (a) not all tools wrap, and (b) it allows you to target your comments more closely to what you want to comment.

> The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this control is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.

Reviewer 3: I don't think this is the foundation for a data pager; a data pager is something you'd use in lieu of this, isn't it?

Author: Yes the thought is if we need to deliver the data management solution as well we would create another control (or create something that inherits from the layout views) that includes the UI component and the data management all in one solution.

> The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this control is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.

Reviewer 3: ```suggestion
The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and WinUI that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in WinUI would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through [XAML Islands](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands), it may also provide an opportunity to benefit WPF and WinForms developers working with DataGrid or ListView. The scope of this control is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.
```

## Description

> <!-- Use this section to provide a brief description of the feature.

Reviewer 3: Could you pull out these template comments?

For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

> A pager control is a UI component that provides a standard interaction for pagination with a layout view (ListView, GridView, ItemsRepeater, DataGrid, etc.). A pager control is independent of the layout views and does not control any of the data being displayed in the layout view. 

Reviewer 3: This isn't a conceptual design guideline document, so we should say throughout 'PagerControl' rather than 'pager control'.

**Important APIs:** [PagerControl class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.pagercontrol)

This control can be configured to use one of three different display modes shown below:

### ComboBox Mode
![](images/pager_control_combo.jpg)

### NumberBox Mode
![](images/pager_control_number.jpg)

### NumericalButtonPanel Mode 
![](images/pager_control_button.jpg)

## Is this the right control? 

Use a **PagerControl** to create a navigation interface for an end user to page through content instead of scrolling or panning. 

> Scrolling through all content in a single view can make it harder for the end user to find specific information and it can lead to performance issues when the content needs to be refreshed. **PagerControl** can be used as an alternate solution so the user can consume a subset of the content without needing to scroll. They can then page through the rest of the content to find what they are looking for. 

Reviewer 3: When should I use this vs ScrollViewer? This text makes it sound like I should never use a ScrollViewer, or there's some threshold number of items/pixels where I should transition to paging?

Reviewer 3: I think the primary scenario for this control is data virtualization; you're trying to show 1000 items, but you don't want to put that load on the server/network in the typical case that the user only looks at the first 20.

> ## Examples

Reviewer 3: These are good examples of the control in isolation, but missing a key case: how do I actually use this with a list?

Author: I was planning on adding that to the XAML controls gallery as a full solution but I can add it to the guidance page as well.

### ComboBox Display Mode

![](images/pager-control-combobox.png)

### Pager Control Editable NumberBox Display Mode 

![](images/pager-control-editable-textbox.png)

### Pager Control Numerical Button Panel Display Mode

![](images/pager-control-numerical-button-panel.png)

A pager control does not require you to provide a max number of pages to the control. If the max number of pages is unknown, here is what each display mode will look like. 

### ComboBox Display Mode

(add comps)

### Pager Control Editable NumberBox Display Mode 

(add comps)

### Pager Control Numerical Button Panel Display Mode

(add comps)

You can choose to customize the look of a pager control by changing the visibility of the first, previous, next, and last buttons. They can be configured to always be visible, hidden, or hidden when the selected index is the first or last page in the range. When the last page is selected, the next and last buttons will be disabled and same when the first page is selected for the first and previous page. The prefix and suffix text can also be customized to display a string you provide. 

## Creating a pager control 

> Here is the XAML for how to add a pager control to your application using the auto display mode. Auto will choose to use the combo box mode if the NumberOfPages property is less than 11 and will show the number box mode if it is greater than 11. In this example, combo box will be chosen. 

Reviewer 3: And if it's exactly 11 it will crash? :)

Reviewer: This also came up during a PR review of the protoype: https://github.com/microsoft/microsoft-ui-xaml/pull/3239#discussion_r483257869

I thought @MarissaMatt updated the proposal, maybe the update wasn't pushed yet?

XAML
```XAML
<Grid>

>         <!-- DataGrid Layout -->

Reviewer 2: This and the comments in samples below seem out of place. We should remove these comments.

        <controls:PagerControl x:Name="MainPagerControl"

>                 DisplayMode="Auto"

Reviewer 2: Several of these properties are set to the default. We should remove them since people tend to copy from the docs.

                DisplayMode="NumericalButtonPanel"
	        NumberOfPages="10"
                FirstButtonVisibility="None"
                PreviousButtonVisibility="AlwaysVisible"
                NextButtonVisibility="AlwaysVisible"
                LastButtonVisibility="None"
	        EllipsisShowFirstAndLast="True">
	</controls:PagerControl>
</Grid>
```

### Customize edge buttons

You can customize the visibility of each of the edge buttons (the first, previous, next, and last) by setting their visibility properties. In this example, the first and previous buttons will always be visible while the next button will be hidden when the last page is selected. The last button will be hidden. By default, all buttons will be set to AlwaysVisible.

XAML
```XAML
<Grid>
        <!-- ListView Layout -->
        <controls:PagerControl x:Name="MainPagerControl"
                DisplayMode="ComboBox"
                NumberOfPages="10"
                FirstButtonVisibility="AlwaysVisible"
                PreviousButtonVisibility="AlwaysVisible"
                NextButtonVisibility="HiddenOnEdge"
                LastButtonVisibility="None"

>                 Prefix="Page"

Reviewer 3: I think these two properties are named PrefixText/SuffixText

>                 Prefix="Page"

Reviewer 2: same here. remove ones which match the default values.

                Suffix="of 10"/>
        </controls:PagerControl>
</Grid>
```

### Handling a page change event

> When a user interacts with the pager control and selects an edge button or page number, the page changed event will be fired. This event will provide the index the user is coming from (previousPage) and the index the user is navigating to (currentPage). Below is a code example of how to handle this event. 

Reviewer 3: What if the user presses next/previous button? Or is this part of "edge buttons"?

Reviewer: I think it would count as edge button here, but that's not really the best name, I am not sure if we need to call out specific ways to interact here. There aren't many ways to interact with the PagerControl and all of them should change the selected index.

XAML
```XAML
<Grid>
        <!-- DataGrid Layout -->
        <controls:PagerControl x:Name="MainPagerControl"
                DisplayMode="Auto"
                NumberOfPages="10"
                FirstButtonVisibility="None"
                PreviousButtonVisibility="AlwaysVisible"
                NextButtonVisibility="AlwaysVisible"
                LastButtonVisibility="None"
                Prefix="Page"
                Suffix="of 10"/>
        </controls:PagerControl>
</Grid>
```
C#
(insert code sample)
```C#
```

## API Notes
| Name | Description| Default | 
|:---:|:---|:---|
| PagerDisplayMode | Enum that contains 4 values (Auto, ComboBox, NumberBox, ButtonPanel) to represent the look of the pager control. When Auto is selected, the display mode will be ComboBox if the NumberOfPages property is less than 11 otherwise it will be NumberBox. | Auto |

> | PagerButtonVisibilityBehavior | Enum that contains 4 values (Auto, AlwaysVisible, HiddenOnEdge, None) that allows the app developer to hide or show the four edge buttons. HiddenOnEdge will remove the appropriate buttons if the selected page is the last or first page. When the last page is selected, the next and last buttons will be disabled and same when the first page is selected for the first and previous page. When Auto is selected, the visibility mode will be AlwaysVisible. | Auto |

Reviewer 2: nit: Is none the same as AlwaysHidden ?

Author: Yes.

Author: @ranjeshj  are you suggesting this should be hidden instead of none? I think @MikeHillberg agreed that none is fine.

Reviewer 2: It wasn't very obvious to me but I'll let the API review crew decide the naming.

Reviewer 3: All the precedent, including in this enum, is for "Hidden". (I think just Hidden, not AlwaysHidden.)

| NumberOfPages | Sets the max number of pages the index control will iterate through. The default will represent an infinite page range. | -1 |
| First, Previous, Next, and Last ButtonCommand | Specially handle the button pressed event for when the end user selects the buttons. | N/A |
| First, Previous, Next, and Last Style | Give the developer the option to customize the style by changing the text or glyph for the edge buttons.| N/A |
| ButtonPanelShowFirstAndLast | Boolean to display the ellipses and the first and last index of the numerical button panel display mode. | True |

> | SelectedIndex | The index that is currently selected. It will default to the first index. | 0 |

Reviewer 3: Should this be -1 if the NumberOfPages is -1?

Reviewer: Should PagerControl even allow negative values? Doesn't really make much sense in my opinion, either there are pages (NumberOfPages  >= 1 pages) or there are none (NumberOfPages = 0). I think using 0 as the default is the best behavior then.

> | SelectedIndex | The index that is currently selected. It will default to the first index. | 0 |

Reviewer 3: Should call out here to be clear: is this 0-relative or 1-relative?

> | PrefixText | Developer can change the prefix text for the combobox and number box display mode options. | "Page" |

Reviewer 3: Please add a note on these two properties that they are ignored by the numerical display mode.

| SuffixText | Developer can change the suffix text for the combobox and number box display mode options. | "of (NumberofPages)". If NumberOfPages is infinite, display nothing |
| PagerControlPageChangedEvent | Event that is fired when the end user selects a button from the number panel, any of the 4 directional buttons, or selects an option from the combobox or enters a number in the numberbox. This event will return the index number that the end user selected. If the end user enters a number greter than the number of pages this event will return the last page. If the end user enters a number less than 1 the event will return 1. | N/A |

> ### Design Components and Functionality

Reviewer 3: This isn't something that should go into the API pages, so for this spec template should go in the appendix.

| Ellipsis | * Button, reading "...", used before or after the first/last index to indicate an accessible but omitted range of indexes. <br> * Visibility of the first/last index can be disabled. <br> <br> * Only visible when using button panel as the display mode.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760513-85277c80-fc79-11e8-926e-3453ea29b0a3.png) |
| PrefixText | * Text displayed before the ComboBox or NumberBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| SuffixText | * When a total number of indices (N) is given, this suffix string will appear after the ComboBox indexing component or NumberBox component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

## API Details

```c++ 
enum PagerDisplayMode
{
    Auto,
    ComboBox,
    NumberBox,
    ButtonPanel,
};

enum PagerButtonVisibilityBehavior
{
    Auto,
    AlwaysVisible,
    HiddenOnEdge,

>     None,

Reviewer: I am not sure if I think Auto is the best description for what it actually does. In a lot of cases, Auto actually has some non static behavior, e.g. [NavigationViewPaneDisplayMode.Auto](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.navigationviewpanedisplaymode?view=winui-2.4). Maybe `Default` is a better name?

Reviewer 4: Without having read to deep into this, wouldn't this be a case similar to the [TabViewCloseButtonOverlayMode](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.tabviewclosebuttonoverlaymode?view=winui-2.4) API? If so, since we used used `Auto` there and not `Default`, we should also use `Auto` here.

Reviewer: It would be the same case as the TabViewCloseButtonOverlayMode, however we also have API's that use Auto in a different way, which makes it confusing for developers to consume APIs given that Auto might be the system default that maps to another value or it might be something different.

Reviewer 4: I think we should stay consistent here within the current API system - and apparently there weren't any real concerns to use `Auto` for the `TabViewCloseButtonOverlayMode` API preventing this name to be used - and I remember `Auto` was discussed a bit. So I favor going with `Auto` here, instead of introducing a new naming scheme, introducing even more inconsistency.

Reviewer: That's true. Maybe we should stick with Auto then.

Author: I have considered changing it to default since there is no logic behind what Auto would do in this case but I agree that there should be some sort of consistency across the controls. @StephenLPeters do you have a recommendation based on the other controls?

Author: After talking internally we decided to keep it as Auto since that is what the majority of the other controls use. I know it's not consistent behavior for every control but in this case we are choosing Auto as the consistent term.

Reviewer 3: The guideline is to use 'Auto' for cases where the control decides for you, which is what this does.  It's not required to by dynamic. (Isn't this dynamic, though, if you change NumberOfPages?)

We've violated that guideline a bunch in the past, but we're trying to stick to it on new APIs.

};

runtimeclass PagerControlPageChangedEventArgs
{
    Integer NewIndex{get; };
    Integer PreviousIndex{get; };
};

runtimeclass PagerControl
{
    PagerControl();

>     PagerDisplayMode Display;

Reviewer 2: consider DisplayMode to match other controls like NavigationView

### Localization
* The numbers will follow standard behavior when localized to/from a language that uses a numeral system that differs from Western Arabic. 
* The prefix and suffix text will follow standard mirroring behavior when localized to/from languages that are RTL or LTR. 

## Data and Intellegence Metrics 

Adoption and validation of the control in the community
* Measurement: Use telemetry to determine how many first and third party developers are using the control in preview. 
* Measurement: Count of posts providing feedback to us through Github and Discord. 

Questions that can be answered through telemetry metrics to help focus future features and improvements: 
Which display mode is the most commonly used? 
* Measurement: Count of how many applications include the pager control type broken out by display mode.

How often are developers using this control with an indefinite number of pages?
* Measurement: Count of how many applications include the pager control type and have the NumberOfPages property empty.

## Open Questions 

> * Should the combo box mode be an editable combo box? My worry with making it an editable combo box is the validity of having the number box mode. Would it hurt to have both editable combo box and number box as options?

Reviewer: That's a really good question. I doubt that there is much benefit in having a destinct NumberBox option when the combo box is editable. I think the big question this boils down to is whether users (as in people using software with that control) expect that ComboBox editable or not.

Author: That is an excellent question for me to think about!

> * I use the terms index and page throughout this document and the APIs. The APIs are all index while the documentation is mostly using page. Should I update the APIs to use page instead of index?

Reviewer: I think it is more natural to refer to index instead of page, after all the pager control does not really say "Get this specific page object" but rather "get page #2". Since we are actually referring to the page by number through the API, index seems more natural I think.

Reviewer 3: Agree with "index" (assuming this is zero-relative). A common use case for this control is to visualize skip/take logic, which is also "pagination", but you do zero-relative indexing on it.

> * There are certain APIs that are specific to one of the three display modes and they will be ignored if the display mode chosen does not support those APIs. Is that something that needs to be reconsidered so it does not cause confusion for the developers?

Reviewer: The only way I could think of, on how we would avoid this, is to have a configuration object that get's passed to the PagerControl. There would be three configuration object types, one for reach display mode and each configuration object only holds fields that are relevant for its display mode.

While this would avoid having properties be irrelevant based on the display mode, this would make it definitely harder to use that control since changing a simple property might get more complex than necessary. @ranjeshj @MikeHillberg What are your thoughts on the concept of a configuration object?

Reviewer 3: Agree that a configuration object would be the most likely solution, and agree that it sounds more complicated than the problem it solves. We should call out each of these in the API Notes, maybe put a "note:" in front of each one to make it possible to look at that table and visualize the extent of the issue.

Reviewer: Such a configuration object surely will be new for developers and a bit complicated. Would it be good to try to have some form of "experiment control" where we explore configuration objects further? It's a bit difficult to see how hard/bad the actual API with configuration objects would look like without actually having tried such a pattern out.

* The enum for the edge button visibility has the four values Auto, AlwaysVisible, HiddenOnEdge, and None. Do those names make sense and translate to the intended behavior? Does the enum name make sense?

> * The ButtonPanelShowFirstAndLast property will show or hide both the first and last page in the numerical button panel. Should this API be split in to two so the developer can show or hide the first and last button individually?

Reviewer: I think there is hardly a use case to only show one of the button panels except when you reach either end of the page list. Splitting this up would also increase the API surface and introduce more properties that are not relevant at all times, so I think we should not split this up.

## Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

A pager control is a UI component that can be added to different layout views (ListView, GridView, ItemsRepeater, or DataGrid) to provide a standard UI interaction for pagination.

**Important APIs:** [PagerControl class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.pagercontrol)

This control can be configured to use three different display modes shown below:

### ComboBox Mode
![](images/pager_control_combo.jpg)

### NumberBox Mode
![](images/pager_control_number.jpg)

### ButtonPanel Mode 
![](images/pager_control_button.jpg)

For more info about choosing the right display mode, see the pager control guidance page.

## Is this the right control? 

> Use a **PagerControl** to create a navigation iterface for an end user to page through content instead of scrolling or panning through the content in a single view. Scrolling through all content in a single view can make it harder for the end user to find specific information and it can lead to performance issues if the view needs to constantly update with new data. **PagerControl** can be used as an alternate solution so the user can consume a subset of the content without needing to scroll. They can then page through the rest of the content to find what they are looking for. 

Reviewer 2: nit: interface (typo)

# Pager control

<!-- The purpose of this spec is to describe a new feature and
its APIs that make up a new feature in WinUI. -->

<!-- There are two audiences for the spec. The first are people
that want to evaluate and give feedback on the API, as part of
the submission process.  When it's complete
it will be incorporated into the public documentation at
docs.microsoft.com (http://docs.microsoft.com/uwp/toolkits/winui/).
Hopefully we'll be able to copy it mostly verbatim.
So the second audience is everyone that reads there to learn how
and why to use this API. -->

> ## Background

Reviewer 3: ```suggestion
## Background

A UI pager is a common control that lets a user page through a long list, rather than scroll or pan. A common experience is to provide preview/next buttons, and a way to jump to a specific page.
```

<!-- Use this section to provide background context for the new API(s) 
in this spec. -->

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

> The absence of a standard UI pager control, last seen as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this proposal is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.

Reviewer 3: ```suggestion
The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this proposal is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.
```

Author: Thank you for the cleaner paragraph! I think this best represents the goal we are trying to achieve with this control.

## Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

> Pager control is a customizable pager UI widget with generic events that can be configured to navigate pages for several view controls.

Reviewer 3: ```suggestion
Use PagerControl to provide the user with a way to page through large content. This can be configured to navigate pages for several view controls.
```

**Important APIs:** [PagerControl class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.pagercontrol)

> Pager control can be configured to use a ComboBox, NumberBox, or numeric Button panel as the core pager control.

Reviewer 3: I don't understand this sentence, PagerControl can be configured to use something else as the pager control?

Author: I'm working on re-writing this sentence to make it clearer but I'm trying to convey that this control has different display modes that a developer can configure to best fit their use case.

Reviewer 5: A picture is worth 1000 words 😉

## Is this the right control? 

> Use a **PagerControl** to create an accessible navigation interface for views that have multiple pages to display. 

Reviewer 3: I think the way to understand this control is that it lets you page through content rather than scrolling/panning through content. I'm not sure what it means to have "pages to display". (Where do I get a page from? Does this only work with the the Xaml Page control?)

Author: That's a good point! I think re-writing this to convey the use case (infinite scrolling vs. pagination) would be better. I will work on updating this.

## Examples

> <!-- Use this section to explain the features of the API, showing

Reviewer 3: Can delete all these comments from the template

example code with each description. The general format is: 
  feature explanation,
  example code
  feature explanation,
  example code
  etc.-->
  
<!-- Code samples should be in C# and/or C++/WinRT -->

<!-- As an example of this section, see the Examples section for the PasswordBox control 
(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->

> A pager control can have several configurations, including these notable ones. 

Reviewer 3: Should use "PagerControl" when we're talking about the control, and "pager control" when we're talking about the abstract UI concept.

> ### ComboBox Pager Control

Reviewer 3: Took me a minute to figure out what it means to be a ComboBox pager control. I think we should describe this as a PagerControl with a ComboBox for picking pages. It's easiest to think about this as a mode of the PagerControl, not a different kind of PagerControl.

Reviewer 3: ```suggestion
### ComboBox Pager Control

This example shows content in a GridView that is navigated using a PagerControl below the GridView. The PagerControl is configured by the `DisplayMode` property to have a ComboBox for the page selector.

```

> ![](images/pager-control-combobox.png)

Reviewer 3: Why is there a scroll bar in this picture? When I use the PagerControl, doesn't that make the ScrollBar not happen?

Reviewer: I think the idea is that we limit the number of item to present the user at once, for example only 20 results from a search. Currently either you would need to implement a pager control yourself or potentially scroll through hundred or thousand of items. 

With the pager control, you would be able to just show `n` items reducing the performance overhead of a search (you would only need to load the first `n` items, which could have an image for example, where loading all at once would be too costly).

Reviewer 3: I see. I think my expectations are misaligned with the goals for this control, and it's flavored my comments. For large content I think the user experience should be scrolling (panning) or paginated, but not both.

Reviewer 6: > I think the idea is that we limit the number of item to present the user at once, for example only 20 results from a search. Currently either you would need to implement a pager control yourself or potentially scroll through hundred or thousand of items.
> 
> With the pager control, you would be able to just show `n` items reducing the performance overhead of a search (you would only need to load the first `n` items, which could have an image for example, where loading all at once would be too costly).

In many forums (which make heavy use of pagination) you can sometimes find an option for how many items to view on each page.  Increase the number of items, decreases the number of pages.

This control is detached from the view with the contents in - so this is probably something outside of the scope of this control, but there should be as much helper code available to determine the page count, from the item count in the list.

Reviewer: But wouldn't a purely paginated approach result in the number of pages essentially changing every time the pager control gets resized? So basically, the page that is currently displayed would need to change every time the app gets resized and update the number of pages as a result of that.

If the PagerControl would be agnostic of the items, essentially like the ItemsRepeater and its layouts, this would actually would be a very interesting control, though it might be confusing for users.

Reviewer: > In many forums (which make heavy use of pagination) you can sometimes find an option for how many items to view on each page. Increase the number of items, decreases the number of pages.

> This control is detached from the view with the contents in - so this is probably something outside of the scope of this control, but there should be as much helper code available to determine the page count, from the item count in the list.

Unless then number of items per page is variable, essentially the number of pages would be given by: 
`NumberOfPages = roundUp(Items/ItemsPerPage)`

I don't think that there would be much a helper could do, since the developer would need to provide a way to determine the number of items per page, and if the developer already is there, the helper wouldn't help much more I think.

Reviewer 3: It does change the number of pages as you resize the control, but I think that's a good thing, like the Read Mode in Word. Actually that's a good prototype we could look at for the experience.

Having both scrolling and pagination at the same time is indeed common on web pages (like the Issues list on github.com), but I've always considered that to be a UX sacrifice made necessary because it's easier to implement.

Reviewer: What about the following scenario: You have the 6th page of 10 open and resize the app (and thus the pager control) to show more items. How would this reflect in the index of the current page and it's content? After all, if we start at page 1 with e.g. 10 items, the 6th page the way it was previously presented could potentially now be out of date and not present the items like it should be presented. Should the items on screen change every time the user resizes the app? Would this be expected if some items suddenly disappear while others might suddenly be appear onscreen?

Reviewer 3: I have actually prototyped this before. I found the app but can't find the source unfortunately. Actually I prototyped it to prove it was a bad idea, but played with it and liked it :)

In that prototype, which was a big list, when you resize the control it updates the number of pages, and updates the current page as necessary to follow the selected item.

Reviewer: > I have actually prototyped this before. I found the app but can't find the source unfortunately. Actually I prototyped it to prove it was a bad idea, but played with it and liked it :)

If you find the source or can share the app, I would love to try that out!

At the end of the day, it's the developers choice to allow scrolling or not inside the pager control content. To avoid any misunderstanding, we could update the screenshot to now shown a scrollbar.

Author: I agree I can update the screenshot to not include the scroll bar for simplicity. @chingucoding are you suggesting that list view and other layout views should have the ability to have both scrolling and pagination? One thing we are considering is adding a property to the layout views to choose between scrolling or paging but the developer would have to choose one or the other.

Reviewer: That's an interesting idea. How would those control behave when the viewport gets smaller or larger? Either you allow scroll or you would need to change the number of items per page, and as a result of that, the current page number.

I think the bigger question here would be how the PagerControl should work. Would you give a control with huge height/width and the PagerControl splits it up (similar to ScrollViewer which takes care of scrolling and the scrollbar) or would the developer need to take care of paginating it's data?
To provide the first case, we would need to find an answer to the first question in this comment.
The second case would be easier for us to define and develop, but would have the disadvantage of being more difficult to use for the developer.

Reviewer 2: When the size changes and we switch pages, we could possibly ensure that the current anchor (some item in the view) still remains in view after the resize even if the page numbers change. Giving a single number of items per page is much simpler but that means avoiding the scrollbar is harder. As already discussed above if we have variable sized items each page might have different number of items and we would need to measure all items ahead to know which items fall in a specific page. We could expose a list of ranges if we wanted that would tell the pager what items fall in each page, but that sounds hard to calculate without breaking virtualization or knowing more about the data/item sizes upfront.

XAML
```XAML
<GridView x:Name="gridView1" ... />

> <controls:PagerControl x:Name="MainPagerControl"

Reviewer 3: Is this the second row of a Grid? Should show that to make it more evident (there is actually a </Grid> below).

        DisplayMode="ComboBox"

> 	NumberOfPages="10"

Reviewer 3: How does an app know how to set this? It determines on the amount of content and the control's actual size, doesn't it?

Author: Are you referring to the NumberOfPages property?

Reviewer 3: Right. As an app, I don't know how many pages there should be; it depends on the size of the list etc. I could see this being a read-only property, but I can't think of a case where an app could or should set it.

Reviewer: It doesn't necessarily depend on the size of the list. For example, you could always present 10 items per page regardless of the available viewport. See [my comment here](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/88#discussion_r451089115) for more context.

Author: This is a property that the developer will set so the pager control will not how many buttons/indices to navigate through. Since this is just a UI component, the pager control will not know anything about the data source or layout view in order to determine how many pages there should be.

Reviewer 2: I think the way it is currently, the app would have to set this based on number of items, the layout it is using and the size of each item since the list and pager are decoupled.

Author: Yes, @Ranjesh is correct that is how it currently works. The developer would need to determine how many pages there should be based on the layout and content being displayed so it would need to be set as well.

	EllipsisMaxAfter="1"/>
</Grid>
```

### Editable NumberBox Pager Control

![](images/pager-control-editable-textbox.png)

XAML
```XAML
<GridView x:Name="gridView1" ... />
<controls:PagerControl x:Name="MainPagerControl"
        DisplayMode="NumberBox"
        NumberOfPages="5"
	FirstButtonVisibility="None"
        PreviousButtonVisibility="AlwaysVisible"
        NextButtonVisibility="AlwaysVisible"
        LastButtonVisibility="None"
	FirstButtonCommand="FirstButtonPressedEvent"
        PreviousButtonCommand="PreviousButtonPressedEvent"
        NextButtonCommand="NextButtonPressedEvent"
        LastButtonCommand="LastButtonPressedEvent"
        PagerInputCommand="PagerInputEvent"
	Suffix="of 5"/>
</Grid>
```

## Adding a PagerControl to Your Application

Here is an example for how to add the PagerControl to your application. It can be added on the page as shown below, or can be used in the template of the layout view you prefer to use. PagerControl is added to the template for DataGrid and ItemsRepeater in WinUI 2.x and will be added to ListView and GridView in WinUI 3. 

(Insert XAML and code behind example here. This will also be added to the XAML controls gallery.)

> ## Remarks

Reviewer 3: I'm working on removing this section from the template. If there are remarks to be made about APIs, they can go in the API Notes section.

<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->

> ## API Notes

Reviewer 3: There should at least be one-line descriptions for any non-obvious APIs here. The one-liners become the IntelliSense text.

Author: I will try and commit the entire one-line description section together so you can see all of them at one time :)

<!-- Option 1: Give a one or two line description of each API (type
and member), or at least the ones that aren't obvious
from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

> ### Visual Components

Author: The visual components are all of the glyphs. I think this was originally meant to just be an area to give descriptions for design functionality so I can change the name to something that makes more sense.

 | Component |  Notes |
|:---:|:---|

> | DisplayMode | * Used to set a button panel, editable ComboBox (default), or NumberBox, as the indexing component. <br> * When set to be a button panel, the number of visible indices can be specified. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760354-23671280-fc79-11e8-804b-71b5b84a80f5.png) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ![image](https://user-images.githubusercontent.com/16964652/49760373-2eba3e00-fc79-11e8-8304-12d7b6ee2c89.png) |

Author: Would it still be appropriate to have this structure in the API notes or appendix section since that content will most likely not transfer to DMC?

| PrefixText | * Text displayed before the editable ComboBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| NumberOfPages | * When a total number of indices (N) is given, this suffix string will appear after the editable ComboBox indexing component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 
enum PagerDisplayMode
{

>     Auto,

Reviewer: Is auto just a default that maps to one of the other three modes or does this behave special compared to the others?

Author: Yes, auto is the default that maps to the combo box display mode.

    ComboBox,
    NumberBox,
    ButtonPanel,
};

> enum ButtonVisibilityMode

Reviewer 3: We need a more specific name. Also, "Mode" doesn't appear to provide value to this name (that word usually doesn't).  PagerButtonVisibility?

Reviewer: When reading "xVisibility" I would expect either "visible/true" or "collapsed/false". Maybe it could be called `PagerButtonVisibilityBehavior` since "AlwaysVisible" or "HiddenOnLast" are mostly behaviors.

{
    Auto,
    AlwaysVisible,

>     HiddenOnLast,

Reviewer: How would this behave?

Author: That is a great question and I will update the API notes section tp make that clear. But the control is the small bar at the bottom that contains the arrows and numbers. The goal is to make this flexible enough that a developer could add the control to the top or the bottom of a layout control.

    None,
};

runtimeclass PagerControlPageChangedEventArgs
{

>     Integer CurrentPage{get; }:

Reviewer 3: This would be better on the control than the args, and we don't need both. So this should be an empty class.

}

> runtimeclass PagerControl

Author: no that was a mistake. I will fix that!

{
    PagerControl();

    PagerDisplayMode Display;
    

>     Integer NumberOfPages;

Reviewer 3: Do we need a change event for this?

Reviewer: The question is, what would you do when that changes? Since this is not something that get's changed from a user interaction and the developer needs to change that value, is there a benefit to providing an event for that?
Especially if that will be a DP, there is limited benefit to providing an event here I think.

Reviewer 3: Sounds good, thanks.

    
    ButtonVisibilityMode FirstButtonVisibility;
    ButtonVisibilityMode PreviousButtonVisibility;
    ButtonVisibilityMode NextButtonVisibility;
    ButtonVisibilityMode LastButtonVisibility;

    FirstButtonCommand="FirstButtonPressedEvent"
    PreviousButtonCommand="PreviousButtonPressedEvent"
    NextButtonCommand="NextButtonPressedEvent"
    LastButtonCommand="LastButtonPressedEvent"
    PagerInputCommand="PagerInputEvent"
    

>     IconSource FirstButtonGlyph;

Reviewer 3: A glyph is always a string today. I think all of our other IconSource properties have an IconSource suffix

Author: We decided to remove these and just keep the button style.

    IconSource PreviousButtonGlyph;
    IconSource NextButtonGlyph;
    IconSource LastButtonGlyph;
    
    String FirstButtonText;
    String PreviousButtonText;
    String NextButtonText;

>     String LastButtonText;

Reviewer: Should we provide localized default texts for those?

Reviewer 2: could we expose the button styles as properties and avoid exposing individual properties ?

Author: Yes! I agree with this approach and we should only expose visibility and the button styles. If we find that developers want to customize the glyphs and text we can expose them later on.

    Windows.UI.Xaml.Input.ICommand FirstButtonCommand;
    Windows.UI.Xaml.Input.ICommand PreviousButtonCommand;
    Windows.UI.Xaml.Input.ICommand NextButtonCommand;
    Windows.UI.Xaml.Input.ICommand LastButtonCommand;
    Windows.UI.Xaml.Input.ICommand PagerInputCommand;
    
    Windows.UI.Xaml.Style FirstButtonStyle;
    Windows.UI.Xaml.Style PreviousButtonStyle;
    Windows.UI.Xaml.Style NextButtonStyle;
    Windows.UI.Xaml.Style LastButtonStyle;
    
    

>     Integer NumberOfIndicesShowing;

Reviewer 3: Why is this necessary?

Author: This is not necessary and I am taking it out of the spec.

>     Boolean EllipsisEnabled;

Reviewer 3: Needs an "Are" prefix to avoid sounding like an event.

>     Boolean EllipsisShowFirstAndLast;

Reviewer 3: Not clear what these Ellipsis properties are?

Reviewer: I think they are for the following ellipsis:

![image](https://user-images.githubusercontent.com/16122379/86832988-33cfb100-c099-11ea-8411-3e0f4e65cce5.png)

Reviewer 3: Looks like there's a description in the API Notes now too. Are these properties ignored when it's a ComboBox?

Author: Yes the ellipses properties are ignored when the developer chooses combo box or number box as the display mode. I can definitely try and find a better way to organize the properties for the different modes but I am not sure if there is a better way to generalize the properties without having subclasses or different classes for each of the three modes.

    Integer EllipsisMaxBefore;
    Integer EllipsisMaxAfter;
    Integer SelectedIndex;
    
    String PrefixText;

>     String SuffixText;

Reviewer: Same here, should we provide localized default texts for those?

Reviewer 2: Yes if these have default text, that should be localized..

Author: Yes the default text should be in the API notes section. PrefixText = "Page" and SuffixText = "of (NumberOfPages)". If NumberOfPages is blank, the SuffixText will be empty.

* Measurement: Count of posts providing feedback to us through Github and Discord. 

Questions that can be answered through telemetry metrics to help focus future features and improvements: 

Which display mode is the most commonly used? 
* Measurement: Count of how many applications include the pager control type broken out by display mode.

Which layout view is this control being used with the most?  
 * Help drive improvements for keyboarding and to detemine which layout views we should focus on adding pager control to first. I'm not sure if this is something that can be measured. 

How often are developers using this control with an indefinite number of pages?
* Measurement: Count of how many applications include the pager control type and have the NumberOfPages property empty. 

## Open Questions 

* I need help understanding how to choose what automation control type could be used for a control like this. Can it be a combination of control types?
* Can the default behavior for combo box and number box be added to pager control? Does anything special need to happen in order to get that functionality?

> * I'm not sure what narrator should announce when focus is on the control. 

Reviewer: I think it should announce the UIA name given by the developer (similiar to NavigationView I think). That way, if there is a meaningful name, the developer can provide it.

Author: That is a really good suggestion! I was debating between saying "Pager" or something a little more clever as the default but I like your suggestion about the developer giving it a name.

# Pager control

## Background

A UI pager is a common control that lets a user page through content, rather than scroll or pan. A common experience is to provide preview/next buttons, and a way to jump to a specific page.

The absence of a standard WinUI pager control, last seen in Xaml as DataPager in Silverlight, has been a pain point in both WPF and WinUI that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in WinUI would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through [XAML Islands](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands), it may also provide an opportunity to benefit WPF and WinForms developers working with DataGrid or ListView. The scope of this control is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.

> ## Description

Reviewer 3: Nit: Background, Description etc are usually level1 headings.

A pager control does not require you to provide a max number of pages to the control. If the max number of pages is unknown, here is what each display mode will look like. 

### ComboBox Display Mode

(add comps)

### Pager Control Editable NumberBox Display Mode 

(add comps)

### Pager Control Numerical Button Panel Display Mode

(add comps)

You can choose to customize the look of a pager control by changing the visibility of the first, previous, next, and last buttons. They can be configured to always be visible, hidden, or hidden when the selected index is the first or last page in the range. When the last page is selected, the next and last buttons will be disabled and same when the first page is selected for the first and previous page. The prefix and suffix text can also be customized to display a string you provide. 

## Creating a pager control 

Here is the XAML for how to add a pager control to your application using the auto display mode. Auto will choose to use the combo box mode if the NumberOfPages property is less than 11 and will show the number box mode if it is greater than 11. In this example, combo box will be chosen. 

XAML
```XAML
<Grid>

>         <controls:DataGrid x:Name="dataGrid" 

Reviewer 3: How does putting a PagerControl next to a DataGrid cause the paging behavior? Is DataGrid being updated to recognize that there's a pager here?

                Height="600" Margin="12"
                AutoGenerateColumns="True"
                ItemsSource="{x:Bind MyViewModel.Items}" />  
        <muxc:PagerControl x:Name="mainPagerControl"

>                 NumberOfPages="10"

Reviewer 3: This doesn't look like a real scenario; how do I know to set the page count to 10? If there are 10,000 items in the DataGrid, this will set up each page to have 1000 items? Shouldn't the page count be a function of the items? Also, is this really a _max_ page count, or will it split 5 rows into 10 pages?

When a user interacts with the pager control and selects an edge button or page number, the page changed event will fire. This event will provide the index the user is coming from (PreviousPageIndex) and the index the user is navigating to (NewPageIndex). 

XAML
```XAML
<Page
    x:Class="WindowsGallaryApp.Samples.DataGridSamplePage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:WindowsGallaryApp.Samples"
    xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
    mc:Ignorable="d"
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Page.Resources>

>         <CollectionViewSource x:Name="imagesCVS" />

Reviewer 3: Is the CVS providing any value here?

Author: No it is not providing any value so I can remove it.

                   Width="190" Height="130" />
        </DataTemplate>
    </Page.Resources>

    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"/>
            <RowDefinition Height="10*"/>
            <RowDefinition Height="Auto"/>
        </Grid.RowDefinitions>
        <TextBlock Grid.Row="0"
                   Text="Image Gallery - DataGrid"
                   Margin="10,0,0,0"
                   Style="{ThemeResource TitleTextBlockStyle}"/>
        <controls:DataGrid Grid.Row="1" 
                           ItemsSource="{x:Bind imagesCVS.View, Mode=OneWay}"
                           AutoGenerateColumns="True"
                           AutoGeneratingColumn="DataGrid_AutoGeneratingColumn"/>
        <data:Pager x:Name="Pager" Grid.Row="2" PageChanged="Pager_PageChanged" PagerDisplayMode="ButtonPanel"/>
    </Grid>
</Page>
```

C#
```C#
        private List<CustomImage> imageList;
        private int imagesPerPage = 25;

        public DataGridSamplePage()
        {
                this.InitializeComponent();
                Loaded += OnPageLoaded;
        }

        private async void OnPageLoaded(object sender, RoutedEventArgs e)
        {

>                 imageList = await GetImageList();

Reviewer 3: Per an earlier comment, I think the primary scenario for this control is data virtualization. Towards that end, to make this sample more realistic, it shouldn't show the images all being pre-fetched; it should show a call to fetch images when the page changes. And then to continue the realism, it needs to show an indeterminate progress bar during the async call to fetch the images from a server?

Author: @MikeHillberg in the Pager_PageChanged method we do call GetPageImages to update the images being shown when the page number changes. OnPageLoaded is only called on the initial load. Am I understanding your comment correctly? Are you saying that the imageList = await GetImageList(); should not be included and the images should be retrieved adhoc?

Reviewer 2: Agree. We could just set the source to a set of images fetched from somewhere based on the page number. That could be a web call for example.

Reviewer 3: @MarissaMatt - Right, in PageChanged its extracting some images. But in PageLoaded it's getting all of the images ready, probably downloading them all from a server. Pre-downloading all the images like that defeat the purpose of the pager, you might as well put them in a ScrollViewer.

Reviewer 2: I updated the sample in the previous comment to do an async call with just the images fetched for each page.

| Ellipsis | * Button, reading "...", used before or after the first/last index to indicate an accessible but omitted range of indexes. <br> * Visibility of the first/last index can be disabled. <br> <br> * Only visible when using button panel as the display mode.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760513-85277c80-fc79-11e8-926e-3453ea29b0a3.png) |
| PrefixText | * Text displayed before the ComboBox or NumberBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| SuffixText | * When a total number of indices (N) is given, this suffix string will appear after the ComboBox indexing component or NumberBox component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

# API Details

> ```c++ 

Reviewer: This looks more like IDL than C++.

Author: @MikeHillberg I was following the template for this section. Is there something I should update this to?

Reviewer: This is technically only for GitHub (and other MarkDown render engines) to know what syntax highlighting to use, but it also helps when reading the raw md file to know what language this is written in. If the API details are in (M)IDL format, this should be IDL. I'm a bit surprised that C++ is in the template, usually APIs are done in IDL as we can simply copy paste that into the actual WinUI source code.

Reviewer 3: It's following the template, except that it needs to be IDL. Changing this to Int32 might be the only issue

enum PagerDisplayMode
{
    Auto,
    ComboBox,
    NumberBox,
    ButtonPanel,
};

enum PagerButtonVisibility
{
    Visible,
    HiddenOnEdge,
    Hidden,
};

> runtimeclass PagerControlSelectedIndexChangedEventArgs

Reviewer: The event later in this document references `PagerControlPageChangedEventArgs` so this class should probably also be named like that.

Author: That is definitely something I didn't update when I changed the name above. Fixing that now.

{
    Integer NewPageIndex{get; };

>     Integer PreviousPageIndex{get; };

Reviewer: If this is IDL, this would be Int32.

};

> runtimeclass PagerControl

Reviewer: ```suggestion
unsealed runtimeclass PagerControl : Windows.UI.Xaml.Controls.Control
```

I don't think there is a reason why this should be sealed. Also we need some form of base class that is a DependencyObject, most other WinUI controls user `Control` for this, so we should do this here too.

Author: good catch!

{
    PagerControl();

    PagerDisplayMode DisplayMode;
    
    Integer NumberOfPages;
    
    PagerButtonVisibility FirstButtonVisibility;
    PagerButtonVisibility PreviousButtonVisibility;
    PagerButtonVisibility NextButtonVisibility;
    PagerButtonVisibility LastButtonVisibility;

    FirstButtonCommand="FirstButtonPressedEvent"
    PreviousButtonCommand="PreviousButtonPressedEvent"
    NextButtonCommand="NextButtonPressedEvent"
    LastButtonCommand="LastButtonPressedEvent"

>     PagerInputCommand="PagerInputEvent"

Reviewer: I assume, with this section, you are trying to point which events on which button should execute that specific command right? In that case, maybe having a comment behind the commands is better to convey that, this might be a bit confusing to read.

Author: Yes Mike pointed this out as well so I will remove it from the API list.

        <muxc:Pager x:Name="pager" Grid.Row="1" PageChanged="OnPageChanged" PagerDisplayMode="NumberPanel"/>
    </Grid>
</Page>
```
C#
```CSharp
public sealed partial class DataGridSamplePage : Page
{
        private const int imagesPerPage = 25;

        public DataGridSamplePage()
        {
                this.InitializeComponent();
        }

        private async void OnPageLoaded(object sender, RoutedEventArgs e)
        {
                pager.NumberOfPages = await ImageGenerator.GetNumberOfPages(imagesPerPage);
                dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(0, imagesPerPage);
                progressRing.Visibility = "Collapsed";
                progressRing.IsActive = false;
        }
        private async void OnPageChanged(Pager sender, PageChangedEventArgs args)
        {
                dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(args.CurrentPage, imagesPerPage);
                progressRing.Visibility = "Visible";
                progressRing.IsActive = true; 
        }
}
```

# API Notes
| Name | Description| Default | 
|:---:|:---|:---|

> | PagerControlDisplayMode | Enum that contains 4 values (Auto, ComboBox, NumberBox, ButtonPanel) to represent the look of the pager control. When Auto is selected, the display mode will be ComboBox if the NumberOfPages property is less than 11 otherwise it will be NumberBox. | Auto |

Reviewer 7: ButtonPanel --> NumericalButtonPanel to match above.

Would love to add BreadbrumbButtonPanel here in the future for the "row of dots" style.

{
        private const int imagesPerPage = 25;

        public DataGridSamplePage()
        {
                this.InitializeComponent();
        }

        private async void OnPageLoaded(object sender, RoutedEventArgs e)
        {
                pager.NumberOfPages = await ImageGenerator.GetNumberOfPages(imagesPerPage);
                dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(0, imagesPerPage);
                progressRing.Visibility = "Collapsed";
                progressRing.IsActive = false;
        }
        private async void OnPageChanged(Pager sender, PageChangedEventArgs args)
        {
                dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(args.CurrentPage, imagesPerPage);
                progressRing.Visibility = "Visible";
                progressRing.IsActive = true; 
        }
}
```

# API Notes
| Name | Description| Default | 
|:---:|:---|:---|
| PagerDisplayMode | Enum that contains 4 values (Auto, ComboBox, NumberBox, ButtonPanel) to represent the look of the pager control. When Auto is selected, the display mode will be ComboBox if the NumberOfPages property is less than 11 otherwise it will be NumberBox. | Auto |

> | PagerButtonVisibilityBehavior | Enum that contains 4 values (Visible, HiddenOnEdge, Hidden) that allows the app developer to hide or show the four edge buttons. HiddenOnEdge will remove the appropriate buttons if the selected page is the last or first page. When the last page is selected, the next and last buttons will be disabled and same when the first page is selected for the first and previous page. When Auto is selected, the visibility mode will be AlwaysVisible. | Visible |

Reviewer: There are only 3 values now, we removed `Auto`.

Author: Updated.

| NumberOfPages | Sets the max number of pages the index control will iterate through. The default will represent an infinite page range. | -1 |
| First, Previous, Next, and Last ButtonCommand | Specially handle the button pressed event for when the end user selects the buttons. | N/A |
| First, Previous, Next, and Last Style | Give the developer the option to customize the style by changing the text or glyph for the edge buttons.| N/A |
| ButtonPanelAlwaysShowFirstAndLastPage | Note: This property only applies to the button panel display mode.Boolean to display the ellipses and the first and last index of the numerical button panel display mode. | True |

> | SelectedIndex | The 0 based index that is currently selected. It will default to the first index. | 0 |

Reviewer: The IDL code says that this is SelectedPageIndex, is that correct, and this entry wasn't updated? Or should we stick with SelectedIndex?

Also, just to clarify, when using this control in ComboBox, when the user clicks on entry "2", the PagerControl will report it's SelectedPageIndex as 1 right?

Author: Yes that is correct! I also updated the name so it will say SelectedPageIndex.

| Ellipsis | * Button, reading "...", used before or after the first/last index to indicate an accessible but omitted range of indexes. <br> * Visibility of the first/last index can be disabled. <br> <br> * Only visible when using button panel as the display mode.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760513-85277c80-fc79-11e8-926e-3453ea29b0a3.png) |
| PrefixText | * Text displayed before the ComboBox or NumberBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| SuffixText | * When a total number of indices (N) is given, this suffix string will appear after the ComboBox indexing component or NumberBox component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

# API Details

```c++

> enum PagerDisplayMode

Reviewer: If the control is named PagerControl, would it be better to name this PagerControlButtonVisibility?

The event args are also named PagerControl... so maybe we should do this here too?

Author: @MikeHillberg do you have an opinion on if this should have the word control the API name?

Reviewer 2: I think it makes sense to call it <ControlName>DisplayMode similar to others like NavigationViewDisplayMode. Property names don't need it since we always use it in the context of the control. So the enum would be PagerControlDisplayMode and the property name would be DisplayMode.

Reviewer 3: I would leave off the "Mode" though; that word always feels good to add to API names, but it doesn't really provide any value.

Reviewer 2: I'm confused, @MikeHillberg are you suggesting removing the 'Mode' so it would be 'PagerControlDisplay' and 'Display' ? That sounds weird to me and would be inconsistent with what we already have in other places like NavigationView.

Reviewer 3: You're right, lots of precedent for ...DisplayMode. Sounds good to me too.

Author: Updated to add control to visibility behavior and display mode.

{
    Auto,
    ComboBox,
    NumberBox,
    ButtonPanel,
};

> enum PagerButtonVisibility

Reviewer: If the control is named PagerControl, would it be better to name this PagerControlButtonVisibility?

{
    Visible,
    HiddenOnEdge,
    Hidden,
};

runtimeclass PagerControlSelectedIndexChangedEventArgs
{
    Integer NewPageIndex{get; };
    Integer PreviousPageIndex{get; };
};

unsealed runtimeclass PagerControlTemplateSettings : Windows.UI.Xaml.DependencyObject
{
        PagerControlTemplateSettings(); 

        Windows.UI.Xaml.Controls.ObservableCollection Pages; 
        Windows.UI.Xaml.Controls.ObservableCollection NumberPanelItems;

        static Windows.UI.Xaml.DependencyProperty PagesProperty { get; };
        static Windows.UI.Xaml.DependencyProperty NumberPanelItemsProperty { get; };

}

> runtimeclass PagerControl

Reviewer: Not sure how that snuck in, but now we have two lines defining the PagerControl class.

Author: Updated.

### ComboBox Display Mode

(add comps)

### Pager Control Editable NumberBox Display Mode 

(add comps)

### Pager Control Numerical Button Panel Display Mode

(add comps)

You can choose to customize the look of a pager control by changing the visibility of the first, previous, next, and last buttons. They can be configured to always be visible, hidden, or hidden when the selected index is the first or last page in the range. When the last page is selected, the next and last buttons will be disabled and same when the first page is selected for the first and previous page. The prefix and suffix text can also be customized to display a string you provide. 

## Creating a pager control 

Here is the XAML for how to add a pager control to your application using the auto display mode. Auto will choose to use the combo box mode if the NumberOfPages property is less than 11 and will show the number box mode if it is greater than 11. In this example, combo box will be chosen. 

XAML
```XAML
<Grid>
        <controls:DataGrid x:Name="dataGrid" 
                Height="600" Margin="12"
                AutoGenerateColumns="True"
                ItemsSource="{x:Bind MyViewModel.Items}" />  

>         <muxc:PagerControl x:Name="mainPagerControl"

Reviewer 2: Is it intentional that both the DataGrid and the Pager are in the same row ? We might want to show two rows and have the PagerControl be in the second row.

Author: the pager is grid.row = 2 and data grid is grid.row =1. Is there another way for me to make sure they are not in the same row?

Reviewer 3: ? Neither one of these is setting Grid.Row

Author: @MikeHillberg is that something that happens with the git commits? It has happened to me before with ProgressRing and I'm not sure how to fix it.

Reviewer 3: It's conspiring against you :) I would just re-copy it, maybe there was a merge error.

    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
    mc:Ignorable="d"
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Page.Resources>
        <CollectionViewSource x:Name="imagesCVS" />
        <DataTemplate x:Key="ImageTemplate" x:DataType="data:CustomImage">
            <Image Stretch="UniformToFill" Source="{x:Bind ImageLocation}"
                   Width="190" Height="130" />
        </DataTemplate>
    </Page.Resources>

    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"/>
            <RowDefinition Height="10*"/>
            <RowDefinition Height="Auto"/>
        </Grid.RowDefinitions>
        <TextBlock Grid.Row="0"
                   Text="Image Gallery - DataGrid"
                   Margin="10,0,0,0"
                   Style="{ThemeResource TitleTextBlockStyle}"/>
        <controls:DataGrid Grid.Row="1" 
                           ItemsSource="{x:Bind imagesCVS.View, Mode=OneWay}"
                           AutoGenerateColumns="True"
                           AutoGeneratingColumn="DataGrid_AutoGeneratingColumn"/>

>         <data:Pager x:Name="MainPager" Grid.Row="2" PageChanged="Pager_PageChanged" PagerDisplayMode="ButtonPanel"/>

Reviewer 2: There is only one pager here, so we could just call this "Pager"

    </Grid>
</Page>
```

C#
```C#
        private List<CustomImage> imageList = new List<CustomImage>();
        private int imagesPerPage = 25;

        public DataGridSamplePage()
        {
            this.InitializeComponent();
            imagesCVS.Source = new ObservableCollection<CustomImage>();

>             Loaded += ListViewSamplePage_Loaded;

Reviewer 2: >ListViewSamplePage_Loaded [](start = 22, length = 25)

OnDataGridSamplePageLoaded

            MainPager.NumberOfPages = (int)Math.Ceiling((double)imageList.Count / imagesPerPage);
            UpdateCollection(MainPager.SelectedIndex);
        }

        private void UpdateCollection(int pageIndex = 0)
        {
            List<CustomImage> images = GetPageImages(pageIndex);
            ((ObservableCollection<CustomImage>)imagesCVS.Source)?.Clear();
            foreach (CustomImage c in images)
            {
                ((ObservableCollection<CustomImage>)imagesCVS.Source)?.Add(c);
            };
        }

        private List<CustomImage> GetPageImages(int pageIndex)
        {
            if ((pageIndex + 1) * imagesPerPage > imageList.Count)
            {
                return imageList.GetRange(pageIndex * imagesPerPage, imageList.Count - (pageIndex * imagesPerPage));
            }

            return imageList.GetRange(pageIndex * imagesPerPage, imagesPerPage);
        }

        private void Pager_PageChanged(Pager sender, PageChangedEventArgs args)
        {
            if (imageList.Count == 0)
            {
                return;
            }

            UpdateCollection(args.NewPageIndex);
        }

>         private void DataGrid_AutoGeneratingColumn(object sender, Microsoft.Toolkit.Uwp.UI.Controls.DataGridAutoGeneratingColumnEventArgs e)

Reviewer 2: We could remove this method

Author: Thank you ranjesh this is much cleaner! I want to test it again before we fully publish the documentation but I appreciate the help!

Reviewer 2: Simplified further with async loading of just the items for each page.

<Page
    x:Class="DataGridSamplePage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"
    xmlns:controls="using:Microsoft.UI.Xaml.Controls"
    Loaded="OnPageLoaded">

    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="*"/>
            <RowDefinition Height="auto"/>
        </Grid.RowDefinitions>
        <controls:DataGrid x:Name="dataGrid" AutoGenerateColumns="True"/>
        <local:Pager x:Name="pager" Grid.Row="1" PageChanged="OnPageChanged" PagerDisplayMode="NumberPanel"/>
    </Grid>
</Page>

 public sealed partial class DataGridSamplePage : Page
    {
        private const int imagesPerPage = 25;

        public DataGridSamplePage()
        {
            this.InitializeComponent();
        }
    
        private async void OnPageLoaded(object sender, RoutedEventArgs e)
        {
            pager.NumberOfPages = await ImageGenerator.GetNumberOfPages(imagesPerPage);
            dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(0, imagesPerPage);
        }
        private async void OnPageChanged(Pager sender, PageChangedEventArgs args)
        {
            dataGrid.ItemsSource = await ImageGenerator.GetImagesInPage(args.CurrentPage, imagesPerPage);
        }
    }

## Open Questions 
* It would be cool to measure which layout view is most commonly used with the pager control, is that possible to measure using telemetry?

> * Is there a shortcut we can add to make it easier to move to the next and previous page? Maybe utilize the left and right bumper?

Reviewer: NavigationView does something similiar. When in DisplayMode top, developers can opt in to let the bumpers change the selection. The API for that is: https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.navigationview.shouldernavigationenabled?view=winui-2.4#Microsoft_UI_Xaml_Controls_NavigationView_ShoulderNavigationEnabled

Reviewer 3: This would only work if you have keyboard focus on the control, which I don't think is a common use case. I think this needs to be a feature of the list control.

### UI Automation Patterns 

### Keyboarding
* Pager Control should be added as a navigation landmark. 
* End users should be able to tab to put focus on the control. 
* Left and right arrows will navigate between the items in the control. 
* Home/escape keys should move focus from an individual item back to the entire control.

> Accessibility for Combo Box Only

Reviewer 3: Is this behavior different than the default behavior? If not, we should just refer to that definition rather than duplicating it here.

Author: I am going to remove this from the spec.

* If the end user is focused on the combo box, the user can hit enter to expand the drop down menu.
* If the end user is focused on the combo box and the drop down is expanded, the home/escape key should collapse the drop down menu and have focus remain on the combo box. 
* If the end user has expanded the drop down list, the PgUp and PgDn buttons should move focus to the first and last items in the list.
* Alt + down shortcut should expand the drop down menu.
* Alt + up shortcut should collapse the drop down menu.

Keyboarding for Number Box Only
* Up and down arrows should increase or decrease the value in the box by 1. 
* PgUp and PgDown arrows will increase or decrease the value in the box by x. The developer will be able to set the amount the value will increase or decrease by.  

### Narrator
* When the focus is on the control, narrator will announce "pager control". 
* When the focus is on the first, last, previous, or next button narrator will announce "first page", "last page", "next page", or "previous page". 
        * If the buttons have text properties set by the developer, narrator will announce that text instead of the default announcement. 

> * When the focus is on a number button, narrator will announce "page x". 

Reviewer 3: No "of y"?
